File: data_structure/linked_list.py
class Node:
    def __init__(self, nData, nextN):
        self.nodeData = nData
        self.nextNode = nextN


class LinkedList:
    def __init__(self):
        self.head = None

    # insertion processes
    def insertAtBeginning(self, newData):
        node = Node(newData, self.head)
        self.head = node

    def insertAtEnd(self, newData):
        node = Node(newData, None)
        if self.head is None:
            self.head = node
        else:
            itr = self.head
            while itr is not None:
                if itr.nextNode is None:
                    itr.nextNode = node
                    break
                itr = itr.nextNode

    def insertValues(self, newValues):
        for data in newValues:
            self.insertAtEnd(data)

    def insertAt(self, index, newData):
        if index < 0 or index >= self.getLength():
            raise Exception("Invalid index")
        elif index == 0:
            self.insertAtBeginning(newData)
        else:
            itr = self.head
            count = 0
            while itr is not None:
                if count == index-1:
                    node = Node(newData, itr.nextNode)
                    itr.nextNode = node
                    break

                count += 1
                itr = itr.nextNode

    # removal processes
    def removeAt(self, index):
        if index < 0 or index >= self.getLength():
            raise Exception("Invalid index")
        elif index == 0:
            self.head = self.head.nextNode
        else:
            itr = self.head
            count = 0
            while itr is not None:
                if count == index-1:
                    itr.nextNode = itr.nextNode.nextNode
                    break
                itr = itr.nextNode
                count += 1

    def remove(self, value):
        if self.exists(value):
            itr = self.head
            if itr.nodeData == value:
                self.head = self.head.nextNode
            else:
                while itr is not None:

                    if itr.nextNode.nodeData == value:
                        itr.nextNode = itr.nextNode.nextNode
                        break
                    itr = itr.nextNode
        else:
            raise Exception(f"{value} does not exist in the linked list")

    def exists(self, value) ->  bool:
        
        return True if self.count(value)>0 else False

    def index(self, value) ->  int:
            itr = self.head
            count = 0
            if self.exists(value):
                while itr is not None:
                    if itr.nodeData == value:
                        return count
                    itr = itr.nextNode
                    count+=1
            else:
                raise Exception(f"{value} does not exist in linked list")
    # return length of items in linked list
    def getLength(self) -> int:
        itr = self.head
        count = 0
        while itr is not None:
            count += 1
            itr = itr.nextNode
        return count

    def count(self, value) -> int:
        itr = self.head
        count = 0
        while itr is not None:
            if itr.nodeData == value:
                count += 1
            itr = itr.nextNode
        return count
    # print linked list

    def printLList(self):
        linked_list = ''
        if self.head is None:
            print("linked list is empty")
        else:
            itr = self.head
            while itr is not None:
                if itr.nextNode is None:
                    linked_list += str(itr.nodeData)
                else:
                    linked_list += str(itr.nodeData) + "-->"

                itr = itr.nextNode
            print(linked_list)


if __name__ == "__main__":
    linkedList = LinkedList()

    linkedList.insertAtEnd(8)
    linkedList.insertAtEnd(80)
    linkedList.insertAtEnd(800)
    
    linkedList.insertValues([1, 2, 3, 4])
    linkedList.insertAt(4, 100)
    print(linkedList.index(4))
    print(linkedList.removeAt(8000))

    linkedList.printLList()

File: data_structure/linked_list.java

public class CustomLinkedList {
	public class Node {
		public int data;
		public Node next;

		public Node(int data) {
			this.data = data;
		}
	}

	public Node head = null;

	public void appendData(int val) {
		Node newDataNode = new Node(val);
		if (head == null) {
			head = newDataNode;
		} else {
			Node currentNode = head;
			while (currentNode.next != null) {
				currentNode = currentNode.next;
			}
			currentNode.next = newDataNode;
		}
	}
	public void pop() {
		if (head == null) {
			return;
		} else {
			Node currentNode = head;
			while (currentNode.next.next != null) {
				currentNode = currentNode.next;
			}
			currentNode.next = null;
		}
	}


	public void prependData(int val) {
		Node newDataNode = new Node(val);
		if (head == null) {
			head = newDataNode;
		} else {
			Node currentNode = head;
			head = newDataNode;
			newDataNode.next = currentNode;
		}
	}

	public int length() {
		if (head == null) {
			return 0;
		} else {
			int counter = 0;
			Node currentNode = head;
			while (currentNode != null) {
				counter++;
				currentNode = currentNode.next;
			}
			return counter;
		}
	}

	public boolean isEmpty() {
		if (length() == 0) {
			System.out.println("true");

			return true;
		} else {
			System.out.println("false");

			return false;
		}
	}

	public void printList() {
		StringBuilder linkedList = new StringBuilder();
		Node currentNode = head;
		while (currentNode != null) {
			linkedList.append(currentNode.data + "->");
			currentNode = currentNode.next;
		}
		System.out.println(linkedList);
	}

	public void insertAt(int index, int val) {
		Node newDataNode = new Node(val);
		if (index > length() || index < 0) {
			System.out.println("index error");
		} else if (index == 0) {
			prependData(val);
		} else if (index == length()) {
			appendData(val);
		} else {
			int counter = 1;
			Node currentNode = head;
			while (counter < index) {
				counter++;
				currentNode = currentNode.next;
			}
			newDataNode.next = currentNode.next;
			currentNode.next = newDataNode;
		}

	}
	public void removeAt(int index) {
		if (index >= length() || index < 0) {
			System.out.println("index error");
		}else if(index == 0) {
			head = head.next;
		}else {
			int counter = 0;
			Node currentNode = head;
			while (counter < index-1) {
				counter++;
				currentNode = currentNode.next;
			}
			currentNode.next = currentNode.next.next;
		}

	}
	public void getValueAt(int index) {
		if (index >= length() || index < 0) {
			System.out.println("index error");
		}else {
			int counter = 0;
			Node currentNode = head;
			while (counter <= index) {
				if(counter == index) {
					System.out.println(currentNode.data);
					return;
				}
				counter++;
				currentNode = currentNode.next;
			}
		}

	}

	public static void main(String[] args) {
		CustomLinkedList customLinkedList = new CustomLinkedList();
		customLinkedList.isEmpty();
		customLinkedList.appendData(50);
		customLinkedList.appendData(1);
		customLinkedList.appendData(32);
		customLinkedList.prependData(2);
		customLinkedList.insertAt(0, 100);
		customLinkedList.insertAt(5, 100);
		customLinkedList.insertAt(3, 21);
		customLinkedList.isEmpty();
		customLinkedList.printList();
		customLinkedList.getValueAt(6);
		customLinkedList.pop();
		customLinkedList.printList();
		customLinkedList.removeAt(0);
		customLinkedList.printList();
		customLinkedList.removeAt(2);
		customLinkedList.printList();

		System.out.println(customLinkedList.length());

	}
}
